### ISSUE TITLE
Vector ++ results are different than in other published benchmarks

### ISSUE THREAD

**Issue by mpilquist** [fosskers/scala-benchmarks#3]
Most notably, Li Haoyi found `Vector ++` to be about twice as fast as `List ++` -- see http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#lists-vs-vectors. It would be nice to understand the discrepancy in results.

Li Haoyi's [summary of `Vector`](http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#vectors-are-ok) also conflicts with the summary here (or at least on Twitter :smile:). If possible, some commentary on the discrepancy would be helpful.

Random hypothesis: given the way `Vector` maintains the most recently accessed 32-array at the root, random access should perform worse than indexed access that looks at relatively adjacent entries. Would be nice to see this confirmed in some tests.

**Comment by fosskers** [comment]
> It would be nice to understand the discrepancy in results.

Definitely. I didn't find `Vector` to be good at much of anything. Seems like the only other thing our two summaries differ on is interation speed - `Vector` loses out there too.

**Comment by fosskers** [comment]
My issue with `Vector` is essentially this - it beats `List` at none of the functional stuff and `Array` at none of the imperative stuff. Okay, so we have a crappy `List` with `O(eC)` random access. What algorithm are we writing that requires random access, but `Array` won't do the job? If holding primitives, `Array` is the only way to go. If classes... yes, what is our algorithm and why can't it be written cleaner with `List` and FP idioms?

**Comment by mpilquist** [comment]
One thing `Vector` solves is the need for an immutable & persistent collection, *in the standard library*, with O(eC) random access. Immutable arrays work nicely if you don't need persistence / can afford copies when you need to build. `mutable.Buffer` or `ArrayBuilder` can help if you can separate modification from access.

What about cache coherency of a `List` that's built up over time? Could building a cons list directly in a loop result in the various cons cells all being relatively adjacent on heap but building it up as part of a larger program result in cons cells being separated? And the resulting cache misses when doing a traversal end up showing list is slower than types that have better cache line behavior? At least with primitive arrays on the JVM, we're guaranteed memory adjacency.

To be clear, I'm not taking the above as truth but rather pointing out that it's a common belief. Putting the belief to the test is great.

**Comment by fosskers** [comment]
Could you expand on the definition of `persistent` in this context? If it's that `Vector` operations share (i.e. references to tails etc are shared across various instances, as `::` with `List` does), then I haven't observed that. I general I'm not convinced that such a collection is useful, `O(eC)` access or not.

Re: the cache. It's true that these benchmarks only test the overhead of the collections themselves in ideal conditions. I can't really make any guarantees about performance under contention, etc.

**Comment by mpilquist** [comment]
Yep, structural sharing. If I recall correctly, `Vector` is implemented with structural sharing but duplicates the path from root. So for example, a vector of size <= 32 results in no structural sharing but one of size 1024 results in sharing of 31 (of 33) nodes upon each change.

Re: cache behavior, that sounds good to me. It's really hard to benchmark stuff and cover the majority of use cases. What got me looking in to these results was a conversation in FS2 Gitter channel as a result of folks interpreting the findings so far as basically "never use `Vector`". I noticed the results in this repo didn't match other benchmarks I've seen and didn't match the general advice of `Vector` being OK at lots of operations but great at none. Anyway, good stuff!

**Comment by fosskers** [comment]
Do you know what operations trigger the sharing? I'd love if `Vector` could pull ahead in *something* in these benchmarks. Otherwise yeah, I would have to take a "never use `Vector`" stance. `List` seems pretty good to me.

**Comment by mpilquist** [comment]
Should be any change -- e.g., given a vector of >= 1024 elements, updating an arbitrary index (e.g. `val w = v.updated(500, 0)`) should result in `w` sharing 31 of 33 nodes with `v`.

**Comment by fosskers** [comment]
Ah, in-place updates (heh, another reason to use Arrays). Too bad this doesn't occur for prepending a la `a +: vect`.
